File: 2092-find-all-people-with-secret/2092-find-all-people-with-secret.java
class Solution {
    public List<Integer> findAllPeople(int n, int[][] meetings, int firstPerson) {
        
        Map<Integer, List<int[]>> graph = new HashMap<>();
        for (int[] meeting : meetings) {
            int person1 = meeting[0];
            int person2 = meeting[1];
            int time    = meeting[2];
            graph.computeIfAbsent(person2, k -> new ArrayList<>()).add(new int[]{time, person1});
            graph.computeIfAbsent(person1, k -> new ArrayList<>()).add(new int[]{time, person2});
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, 0});
        pq.offer(new int[]{0, firstPerson});

        boolean[] visited = new boolean[n];

        while (!pq.isEmpty()) {
            int[] timePerson = pq.poll();
            int time = timePerson[0], person = timePerson[1];
            if (visited[person]) {
                continue;
            }
            visited[person] = true;
            for (int[] nextPersonTime : graph.getOrDefault(person, new ArrayList<>())) {
                int t = nextPersonTime[0], nextPerson = nextPersonTime[1];
                if (!visited[nextPerson] && t >= time) {
                    pq.offer(new int[]{t, nextPerson});
                }
            }
        }
        
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; ++i) {
            if (visited[i]) {
                ans.add(i);
            }
        }
        return ans;
    }
}

File: LeetCode/2092-find-all-people-with-secret/2092-find-all-people-with-secret.py
class Solution:
    def findAllPeople(self, n: int, meetings: List[List[int]], firstPerson: int) -> List[int]:

        graph = collections.defaultdict(list)
        for meeting in meetings:
            person1, person2, time = meeting
            graph[person2].append((time, person1))
            graph[person1].append((time, person2))

        pq = [(0, 0), (0, firstPerson)]
        heapq.heapify(pq)

        visited = [False] * n

        while pq:
            time, person = heapq.heappop(pq)
            if visited[person]:
                continue
            visited[person] = True
            for t, nextPerson in graph.get(person, []):
                if not visited[nextPerson] and t >= time:
                    heapq.heappush(pq, (t, nextPerson))

        ans = []
        for i in range(n):
            if visited[i]:
                ans.append(i)
        return ans